File: backend/app/schemas/influencer.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import date, datetime
from enum import Enum

# Enums for structured data
class SignalType(str, Enum):
    BUY = "BUY"
    SELL = "SELL"
    HOLD = "HOLD"       # Legacy — mapped to WATCH in new prompts
    HEDGE = "HEDGE"
    WATCH = "WATCH"
    CLOSED = "CLOSED"

class TimeframeType(str, Enum):
    SHORT = "SHORT"   # < 1 week
    MID = "MID"       # 1-4 weeks
    LONG = "LONG"     # > 1 month

class SourceType(str, Enum):
    MANUAL = "MANUAL"
    AUTO_THREADS = "AUTO_THREADS"
    AUTO_SUBSTACK = "AUTO_SUBSTACK"

class RecommendationStatus(str, Enum):
    PENDING = "PENDING"    # Awaiting user review (auto-tracked only)
    ACTIVE = "ACTIVE"      # Currently being tracked
    EXPIRED = "EXPIRED"    # Past expiry date
    CLOSED = "CLOSED"      # Manually closed by user

# Influencer models
class InfluencerBase(BaseModel):
    name: str
    platform: Optional[str] = None
    url: Optional[str] = None

class InfluencerCreate(InfluencerBase):
    pass

class Influencer(InfluencerBase):
    id: str
    created_at: datetime

class InfluencerUpdate(BaseModel):
    name: Optional[str] = None
    platform: Optional[str] = None
    url: Optional[str] = None

# Enhanced Recommendation models
class RecommendationBase(BaseModel):
    symbol: str
    signal: SignalType = SignalType.BUY
    timeframe: TimeframeType = TimeframeType.MID
    recommendation_date: date
    entry_price: Optional[float] = None
    target_price: Optional[float] = None
    stop_loss: Optional[float] = None
    expiry_date: Optional[date] = None
    source_url: Optional[str] = None
    note: Optional[str] = None

class RecommendationCreate(RecommendationBase):
    source: SourceType = SourceType.MANUAL

class Recommendation(RecommendationBase):
    id: str
    influencer_id: str
    source: Optional[str] = None  # str instead of enum to tolerate legacy data
    status: RecommendationStatus = RecommendationStatus.ACTIVE
    created_at: datetime
    # Calculated fields (populated by backend)
    current_price: Optional[float] = None
    unrealized_return: Optional[float] = None
    final_return: Optional[float] = None
    hit_target: Optional[bool] = None
    hit_stop_loss: Optional[bool] = None

class RecommendationUpdate(BaseModel):
    symbol: Optional[str] = None
    signal: Optional[SignalType] = None
    timeframe: Optional[TimeframeType] = None
    recommendation_date: Optional[date] = None
    entry_price: Optional[float] = None
    target_price: Optional[float] = None
    stop_loss: Optional[float] = None
    expiry_date: Optional[date] = None
    source: Optional[SourceType] = None
    source_url: Optional[str] = None
    note: Optional[str] = None
    status: Optional[RecommendationStatus] = None

# Stats models for performance tracking
class InfluencerWithStats(Influencer):
    recommendation_count: int = 0
    active_count: int = 0
    expired_count: int = 0
    win_rate: Optional[float] = None  # % of expired recommendations with positive return
    avg_return: Optional[float] = None
    hit_target_rate: Optional[float] = None

class InfluencerPerformance(BaseModel):
    """Detailed performance breakdown by timeframe"""
    influencer_id: str
    name: str
    short_term: Optional[dict] = None  # {count, win_rate, avg_return}
    mid_term: Optional[dict] = None
    long_term: Optional[dict] = None
    overall: Optional[dict] = None

# Pending review model for auto-tracked recommendations
class PendingReview(BaseModel):
    id: str
    influencer_id: str
    influencer_name: str
    source: Optional[str] = None
    source_url: str
    original_content: str
    ai_analysis: dict  # Raw AI output
    suggested_symbol: Optional[str] = None
    suggested_signal: Optional[SignalType] = None
    suggested_timeframe: Optional[TimeframeType] = None
    confidence: Optional[float] = None
    post_date: Optional[date] = None
    created_at: datetime

File: frontend/src/domain/models/Influencer.ts
// Enums for structured data
export type SignalType = 'BUY' | 'SELL' | 'HOLD' | 'HEDGE' | 'WATCH' | 'CLOSED';
export type TimeframeType = 'SHORT' | 'MID' | 'LONG';
export type SourceType = 'MANUAL' | 'AUTO_THREADS' | 'AUTO_SUBSTACK';
export type RecommendationStatus = 'PENDING' | 'ACTIVE' | 'EXPIRED' | 'CLOSED';

export interface Influencer {
  id: string;
  name: string;
  platform?: string;
  url?: string;
  created_at: string;
  recommendation_count?: number;
}

export interface Recommendation {
  id: string;
  influencer_id: string;
  symbol: string;
  signal: SignalType;
  timeframe: TimeframeType;
  recommendation_date: string;
  entry_price: number | null;
  target_price?: number | null;
  stop_loss?: number | null;
  expiry_date?: string | null;
  source: SourceType;
  source_url?: string;
  note?: string;
  status: RecommendationStatus;
  created_at: string;

  // Calculated fields from backend
  current_price?: number;
  unrealized_return?: number;
  final_return?: number;
  hit_target?: boolean;
  hit_stop_loss?: boolean;
}

export interface InfluencerCreate {
  name: string;
  platform?: string;
  url?: string;
}

export interface RecommendationCreate {
  symbol: string;
  signal?: SignalType;
  timeframe?: TimeframeType;
  recommendation_date: string;
  entry_price?: number;
  target_price?: number;
  stop_loss?: number;
  expiry_date?: string;
  source?: SourceType;
  source_url?: string;
  note?: string;
}

export interface InfluencerUpdate {
  name?: string;
  platform?: string;
  url?: string;
}

export interface RecommendationUpdate {
  symbol?: string;
  signal?: SignalType;
  timeframe?: TimeframeType;
  recommendation_date?: string;
  entry_price?: number;
  target_price?: number;
  stop_loss?: number;
  expiry_date?: string;
  source?: SourceType;
  source_url?: string;
  note?: string;
  status?: RecommendationStatus;
}

export interface InfluencerWithStats extends Influencer {
  active_count?: number;
  expired_count?: number;
  win_rate?: number;
  avg_return?: number;
  hit_target_rate?: number;
}

export interface PendingReview {
  id: string;
  influencer_id: string;
  influencer_name: string;
  source: SourceType;
  source_url: string;
  original_content: string;
  ai_analysis: Record<string, unknown>;
  suggested_symbol?: string;
  suggested_signal?: SignalType;
  suggested_timeframe?: TimeframeType;
  confidence?: number;
  created_at: string;
}

// Helper functions
export const getTimeframeLabel = (tf: TimeframeType, t: any): string => {
  switch (tf) {
    case 'SHORT':
      return t('influencers.timeframes.SHORT', { defaultValue: '短期 (<1週)' });
    case 'MID':
      return t('influencers.timeframes.MID', { defaultValue: '中期 (1-4週)' });
    case 'LONG':
      return t('influencers.timeframes.LONG', { defaultValue: '長期 (>1月)' });
    default:
      return tf;
  }
};

export const getSignalLabel = (signal: SignalType, t: any): string => {
  switch (signal) {
    case 'BUY':
      return t('influencers.signals.BUY', { defaultValue: '看多 📈' });
    case 'SELL':
      return t('influencers.signals.SELL', { defaultValue: '看空 📉' });
    case 'HOLD':
      return t('influencers.signals.HOLD', { defaultValue: '觀望 ⏸️' });
    case 'HEDGE':
      return t('influencers.signals.HEDGE', { defaultValue: '避險 🛡️' });
    case 'WATCH':
      return t('influencers.signals.WATCH', { defaultValue: '觀察 👀' });
    case 'CLOSED':
      return t('influencers.signals.CLOSED', { defaultValue: '已平倉 ✅' });
    default:
      return signal;
  }
};

export const getStatusLabel = (
  status: RecommendationStatus,
  t: any
): string => {
  switch (status) {
    case 'PENDING':
      return t('influencers.statuses.PENDING', { defaultValue: '待審核' });
    case 'ACTIVE':
      return t('influencers.statuses.ACTIVE', { defaultValue: '追蹤中' });
    case 'EXPIRED':
      return t('influencers.statuses.EXPIRED', { defaultValue: '已到期' });
    case 'CLOSED':
      return t('influencers.statuses.CLOSED', { defaultValue: '已結束' });
    default:
      return status;
  }
};
